proposals/0281-visible-forall.rst

```diff
@@ -1242,6 +1341,33 @@ Alternatives
 
    Instead, we opted to raise the ambiguity error during T2T.
 
+8. We could permit visible ``forall`` in Haskell98-style data declarations,
+   where invisible ``forall`` is already allowed to bind existential variables.
+
+   However, GHC does not support syntactically nested foralls in such
+   declarations. Compare:
+
+   * ``data Flat = forall a b. MkFlat a b``
+   * ``data Nested = forall a. forall b. MkNested a b``
+
+   At the time of writing this proposal, ``Flat`` is a valid declaration
+   whereas ``Nested`` results in a parse error.
+
+   This means that the following declaration would be rejected even if we
+   allowed visible forall::
+
+      data MixedVis = forall a. forall b -> MkMixedVis a b
+
+   A similar limitation in GADTs is addressed by GHC Proposal `#402 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0402-gadt-syntax.rst>`_
+   (accepted, not implemented). In principle, nothing precludes us from
+   extending the syntax of existential quantification in Haskell98-style data
+   declarations in a similar manner, but nothing compels us to do so either.
+
+9. We could say that ``a -> b`` is a view pattern regardless of enabled
+   extensions and thereby make this proposal less fork-like; the problem is
+   that it would create an asymmetry between use sites ``f (MkT (Int -> Bool))``
+   and definition sites ``f (MkT (a -> b)) = ...``.
```

Comment:
I did not understand the asymmetry.  Could you spell it out please?

Perhaps you mean: `(e1 -> e2)` would be allowed in terms but `(p1 -> p2)` would not be allowed in patterns?   Better to say it like that, if that's what you mean.

And yes, that's asymmetric, but that's what happens if you add `ViewPatterns`.  The question is whether the asymmetry changes with other (apparently unrelated) extensions.